kafka-rest/src/main/java/io/confluent/kafkarest/ProducerMetrics.java

```diff
@@ -0,0 +1,395 @@
+import org.apache.kafka.common.MetricNameTemplate;
+import org.apache.kafka.common.metrics.JmxReporter;
+import org.apache.kafka.common.metrics.KafkaMetricsContext;
+import org.apache.kafka.common.metrics.MeasurableStat;
+import org.apache.kafka.common.metrics.MetricConfig;
+import org.apache.kafka.common.metrics.Metrics;
+import org.apache.kafka.common.metrics.MetricsContext;
+import org.apache.kafka.common.metrics.MetricsReporter;
+import org.apache.kafka.common.metrics.Sensor;
+import org.apache.kafka.common.metrics.stats.Avg;
+import org.apache.kafka.common.metrics.stats.CumulativeCount;
+import org.apache.kafka.common.metrics.stats.CumulativeSum;
+import org.apache.kafka.common.metrics.stats.Max;
+import org.apache.kafka.common.metrics.stats.Rate;
+import org.apache.kafka.common.metrics.stats.WindowedCount;
+import org.apache.kafka.common.metrics.stats.WindowedSum;
+import org.apache.kafka.common.utils.Time;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ProducerMetrics {
+  public static final String JMX_PREFIX = "kafka.rest";
+
+  private static final Logger log = LoggerFactory.getLogger(ProducerMetrics.class);
+
+  private final Metrics metrics;
+  private final ConcurrentMap<BeanCoordinate, ProduceMetricMBean> beansByCoordinate =
+      new ConcurrentHashMap<>();
+
+  public ProducerMetrics(Time time) {
+    this.metrics = new MetricsBuilder(JMX_PREFIX).withTime(time).build();
+    setupMetricBeans();
+    log.info("Successfully registered kafka-rest produce metrics with JMX");
+  }
+
+  /**
+   * Get or create a {@link ProduceMetricMBean} with the specified name and the given tags. Each
+   * group is uniquely identified by the name and tags.
+   *
+   * @param groupName the name of the ReplicatorMetricGroup group; may not be null
+   * @param tags pairs of tag name and values
+   * @return the {@link ProduceMetricMBean} that can be used to create metrics; never null
+   */
+  public ProduceMetricMBean mbean(String groupName, Map<String, String> tags) {
+    BeanCoordinate beanCoordinate = new BeanCoordinate(groupName, tags);
+    beansByCoordinate.putIfAbsent(beanCoordinate, new ProduceMetricMBean(beanCoordinate));
+    return beansByCoordinate.get(beanCoordinate);
+  }
+
+  /** Stop and unregister the metrics from any reporters. */
+  public void stop() {
+    log.info("Unregistering produce metrics with JMX");
+    metrics.close();
+  }
+
+  /** Sets up a {@link ProduceMetricMBean} for each sensor */
+  private void setupMetricBeans() {
+    ProduceMetricMBean metricMBean =
+        mbean(ProducerMetricsRegistry.GROUP_NAME, Collections.emptyMap());
+
+    // Clear out any previous metrics and sensors that might have been present
+    if (metricMBean != null) {
+      metricMBean.close();
+    }
+
+    // request metrics
+    new SensorBuilder(metricMBean, ProducerMetricsRegistry.REQUEST_SENSOR)
+        .addRate(ProducerMetricsRegistry.REQUEST_RATE, ProducerMetricsRegistry.REQUEST_RATE_DOC)
```

Comment:
In counter to what Dimitar has said, why not use all the calculation logic that is built in to the metrics, to save us having to do it ourselves?  We export request rate for the other rest calls, I would propose that produce is no different.  

Is there a harm in adding it here?  (Especially if we don't export in dd-agent to start with, but the info is there if we need it).

Reply:
@dimitarndimitrov the rate metrics are consistent with the other clients (and I believe we even kept them alongside windowedcounts upstream). If we were to remove any of them it should be the windowedcount metrics imo.

Reply:
The reason I don't like the built-in `Rate` metric is that it's not very flexible and it's hard to interpret because it naturally results in double aggregation. It's a sliding window rate (I personally prefer exponential weighting) which in its most simple form just uses some default window and you have to look into `ce-kafka` to find what that window is.

I personally think that metric-wise we need very little more than histogrammed counts and latencies, if we can use a dynamic range histogram, with raw values as a close second (or first, if we can't use good histograms). Everything that can be inferred from that at a later point seems like a redundancy that we want to avoid at runtime.
- The same applies for the rates on the Admin APIs - I actually think we should calculate these from the request counts. My main concern with cumulative vs windowed counts was the potential loss of precision/fidelity if we were to use these for rate calculation, but that didn't turn out to be a problem.

I am very open to changing my mind on this, but if I happen to be pig-headed without even realising that, please do let me know that :D

Reply:
@dimitarndimitrov My worry is that we don't offer any comparable metrics between other clients and the rest producer. All of the Confluent developed clients offer a rate metric defined in this way and I think it is expected. And, if present, should be consistent with the other clients.